### DISCUSSION TITLE
Luhmann-like file-naming scheme

### DISCUSSION THREAD

**Discussion by gour** [zk-org/zk#308] [Ideas]
Hello,

recently I watched the video [series](https://www.thesing-online.de/blog/denote-as-a-zettelkasten) which describes Zettelkasten note-taking by using Emacs' [Denote](https://protesilaos.com/emacs/denote) package. It uses a very interesting file-naming scheme in the `signature` branch allowing one to name one's notes with the “signatures” which do resemble, afaict, the original Luhmann's index-card's naming scheme. This allows to “…be used to establish a sequential relationship between files (e.g. 1, 1a, 1b, 1b1, 1b2, …)”, so I wonder if something like that could be deployed with `zk`?

Denote package itself is very powerful, but I'm simply still not sold on Emacs…not knowing Elisp etc…. :smile:

**Comment by mickael-menu** [comment]
It is not possible out-of-the-box with `zk`. However you could do some scripting to extract the current ID from the filename or the YAML frontmatter and build a new child.

For example, given this note:

```
---
id: 1b
---

# Test
```

You can get the `id` with:

```shell
$ zk list test.md --format json | jq -r .[0].metadata.id
1b
```

Then use a [`zk` alias](https://github.com/mickael-menu/zk/blob/main/docs/config-alias.md) to add a new command such as `zk branch`.

**Comment by gour** [comment]
> It is not possible out-of-the-box with `zk`. However you could do some scripting to extract the current ID from the filename or the YAML frontmatter and build a new child.

Zk is, again, catching my interest...I've thought that instead of using separate metadata (`id`), I could simply use Luhmann's naming scheme as one of the tags, since filtering by them is nicely supported by `zk`. 

What do you think?

**Reply by tjex** [reply]
I think then you may dilute the efficacy of the tagging system for filtering / search narrowing. Tags should be used to find notes that share some attribute(s) but are not necessarily directly related. Your tag panel would become unwieldy. 

My opinion with regards to Luhmann's numbering system, is that it was developed within the limitations of his paper and pen system. The numbering system did help him to think logically and iterate on ideas, but it also served as an organisational system so that he could physically place and find notes. It's in general a contentious issue as to its value: https://zettelkasten.de/folgezettel/

With a digital system (hyperlinks), the ordering of the notes in 'space' within a draw / in a directory is no longer necessary.

A happy medium, which is often recommended, is to simply prepend all your notes with a date stamp. This creates a natural chronological ordering, and when combined with outbound links, allows you to identify which notes in the sequence are parents / children of others.

**Comment by artkpv** [comment]
@mickael-menu Thank you and other contributors for this tool, `zk`. I'm now interested to migrate to this from https://github.com/lervag/wiki.vim . However, I use this Luhmann-like sequencial ids. See below an example. The reason was to make it simple to reference up and down notes, i.e. a note at 0/01.1b for example don't need to contain links to the upper note 0/01.1, nor to the lower notes, like 0/01.b2. This way I can quickly see links (I just call a file explorer from my vim config that focuses at a current buffer file). I wrote two simple scripts to create a new note from current one (similar to your `zk branch` above) and to move / rename notes that tries to maintain the order. The renaming script is not optimal (buggy, runs slowly). 

How do you think I can make this happen with `zk`? Or it is not practical, but then why? It should be just random ids. Or perhaps there is some `zk branch` more powerful then the above? 

```
    > ls 0/  # From my root for notes

    0/01 <title>.md
    0/01.1 <title>.md
    0/01.1b <title>.md
    0/01.1b2 <title>.md
    0/01.2 <title>.md
    0/01.2a <title>.md
    0/01.2a1 <title>.md
    0/01.2b <title>.md
    0/01.2b1 <title>.md
    0/01.2b2 <title>.md
    0/01.3 <title>.md
    # ...
```

**Reply by tjex** [reply]
Implementing this with zk is arguably more practical than other zk tools, given zk's capabilities with templating / scripting / aliases, etc. Any sense of practicality past that is not related to the tool, but to the system of a Folgezettel UID in the first place. The UID is a necessity if you're using a physical Zettelkasten, as you have to know where in the order to _pull a note out to read it and put it back physically so you can find it again_. 

In a digital zettelkasten this isn't needed, as we don't remove notes and place them back manually. So the tradeoffs of managing a Folgezettel UID system weigh more on the tradeoff side, rather than benefits.

**Reply by artkpv** [reply]
> Implementing this with zk is arguably more practical than other zk tools, given zk's capabilities with templating / scripting / aliases, etc. Any sense of practicality past that is not related to the tool, but to the system of a Folgezettel UID in the first place. The UID is a necessity if you're using a physical Zettelkasten, as you have to know where in the order to _pull a note out to read it and put it back physically so you can find it again_.
> 
> In a digital zettelkasten this isn't needed, as we don't remove notes and place them back manually. So the tradeoffs of managing a Folgezettel UID system weigh more on the tradeoff side, rather than benefits.

Thanks. I made decision to use Random IDs as they're described in `zk`, and drop the sequential ids. Though I have many notes with those sequential ids as I wrote above, I hope to transition gradually. I'm now adopting many "hub" or index notes for that purpose.

**Reply by tjex** [reply]
An in between, if you still want sequentiality in general in your filesystem, is to just use something like `filename = "{{sh 'date +%y%m%d%H%M'}}"`. It's a bit long though, but I've grown to be ok with it.

I used to use the four character random ID. The only reason I changed, was after reading how another user found that in a note with many subsequent references (eg in the footer), that the date can help to direct them as to which one to open (many years ago or last month can both be interesting options).
